README.md
WebSocket.IO
============

[![Build Status](https://secure.travis-ci.org/LearnBoost/websocket.io.png)](http://travis-ci.org/LearnBoost/websocket.io)

WebSocket.IO is an abstraction of the websocket server previously used by Socket.IO.
It has the broadest support for websocket protocol/specifications and an API that
allows for interoperability with higher-level frameworks such as
[Engine](http://github.com/learnboost/engine.io),
[Socket.IO](http://github.com/learnboost/socket.io)'s realtime core.

## Features

- Fast
- Minimalistic
  - Offers an integration API for higher-level impls to handle authorization,
    routing, etc
- Widest support of protocols
  - Draft-75
  - Draft-76
  - Draft-00
  - Protocol version 7
  - Protocol version 8
  - Protocol version 13
- Written for Node 0.6

## How to use

### Server

#### (A) Listening on a port

```js
var ws = require('websocket.io')
  , server = ws.listen(3000)

server.on('connection', function (socket) {
  socket.on('message', function () { });
  socket.on('close', function () { });
});
```

#### (B) Intercepting WebSocket requests for a http.Server

```js
var ws = require('websocket.io')
  , http = require('http').createServer().listen(3000)
  , server = ws.attach(http)

server.on('connection', function (socket) {
  socket.on('message', function () { });
  socket.on('close', function () { });
});
```

#### (C) Passing in requests

```js
var ws = require('websocket.io')
  , server = new ws.Server()

server.on('connection', function (socket) {
  socket.send('hi');
});

// …
httpServer.on('upgrade', function (req, socket, head) {
  server.handleUpgrade(req, socket, head);
});
```

### Client-side example

```js
var ws = new WebSocket("ws://host:port/");        

ws.onopen = function() {
 //do something when connection estabilished
};

ws.onmessage = function(message) {
 //do something when message arrives
};

ws.onclose = function() {
 //do something when connection close
};

```

## API

### Server

<hr><br>

#### Top-level

These are exposed by `require('websocket.io')`:

##### Properties

- `version` _(String)_: websocket.io version
- `protocols` _(Object)_: constructors for drafts 75/76/00, protocols 7/8/13.

##### Methods

- `listen`
    - Creates an `http.Server` which listens on the given port and attaches WS
      to it. It returns `501 Not Implemented` for regular http requests.
    - **Parameters**
      - `Number`: port to listen on
      - `Object`: optional options object. See `Server` for options details.
      - `Function`: callback for `listen`
    - **Returns** `Server`
- `attach`
    - Captures `upgrade` requests for a `http.Server`. In other words, makes a regular `http.Server` websocket-compatible.
    - **Parameters**
      - `http.Server`: server to attach to.
      - `Object`: optional, options object. See `Server` for options details.
    - **Returns** `Server`

<hr><br>

#### Server

The main server/manager. _Inherits from EventEmitter_.

##### Properties

- `clients` _(Object)_: table of all connected clients. This property is
  not set when the `Server` option `clientTracking` is `false`.
- `clientsCount` _(Number)_: total count of connected clients. This property is
  not set when the `Server` option `clientTracking` is `false`.

##### Events

- `connection`
    - Fired when a new connection is established. 
    - **Arguments**
      - `Socket`: a Socket object

##### Methods

- **constructor**
    - Initializes the server
    - **Options**
      - path (`String`): if set, server only listens to request for this path
      - clientTracking (`Boolean`): enables client tracking. The
        `Server#clients` property only is available when this is true (`true`)
- `handleUpgrade`

<hr><br>

#### Socket

A representation of a client. _Inherits from EventEmitter_.

##### Properties

- `req` _(http.ServerRequest)_: Request that originated the connection.
- `socket` _(net.Socket)_: Stream that originated the connection.
- `readyState` _(String)_: `opening`, `open`, `closed`.
- `name` _(String)_: `websocket-hixie`, `websocket`.
- `protocolVersion` _(String)_: `hixie-75`, `hixie-76` and `hybi`.

##### Events

- `close`
    - Fired when the connection is closed.
- `message`/`data`
    - Fired when data is received
    - **Arguments**
      - `String`: utf-8 string
- `error`
    - Fired when an error occurs.
    - **Arguments**
      - `Error`: error object

##### Methods

- `send` / `write`
    - Sends a message.
    - **Parameters**
      - `String`: utf-8 string with outgoing data
    - **Returns** `Socket` for chaining
- `close` / `end`
    - Closes the socket
    - **Returns** `Socket` for chaining
- `destroy`
    - Forcibly closes the socket.

## Support

The support channels for `websocket.io` are the same as `socket.io`:

  * irc.freenode.net **#socket.io**
  * [Google Groups](http://groups.google.com/group/socket_io)
  * [Website](http://socket.io)

## Development

To contribute patches, run tests or benchmarks, make sure to clone the
repository:

```
git clone git://github.com/LearnBoost/websocket.io.git
```

Then:

```
cd websocket.io
npm install
```

## Tests

```
$ make test
```

## Credits

WebSocket.IO is possible thanks to the contributions by:

- Einar Otto Stangvik [einaros@gmail.com]
- Arnout Kazemier [info@3rd-eden.com]
- Nico Kaiser [nico@kaiser.me]

## License 

(The MIT License)

Copyright (c) 2011 Guillermo Rauch &lt;guillermo@learnboost.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.2.2 / 2017-01-02
==================

  * Updated EventEmitter to work with node v7
  * Allow mocha tests to run on ipv6 machines

0.2.1 / 2012-07-03
==================

  * Bumped ws
  * Emit headers event for drafts protocol
  * Added `data` event to `Socket`
  * Improves `clientTracking` memory efficiency [ruxkor]

0.2.0 / 2012-06-19
==================

  * Check if socket is writable before finishing handshake
  * Implement readyState and async connection opening
  * Check for readyState before writing
  * Fix wrong interface for hybi protocol
  * Remove logger, use debug for logging
  * Disable client tracking in `ws` (hybi)
  * client tracking -> clientTracking
  * Use `ws` server and parsers for hybi protocols
  * Handle socket "timeout" event
  * Add missing Sec-WebSocket-Protocol header to hybi protocols

0.1.6 / 2012-02-27
==================

  * Fix hybi-16 on node 0.4

0.1.5 / 2012-02-10
==================

  * Bumped ws

0.1.4 / 2012-01-04
==================

  * `close` event fires in a different tick after calling Socket#close

0.1.3 / 2012-01-02
==================

  * Bumped ws
  * Cleaned up tests
  * Cleaned up docs

0.1.2 / 2011-12-12
==================

  * Updated ws
  * Added 0.4 compatibility.

0.1.1 / 2011-11-27
==================

  * Added echo example.
  * Fixed/improved parser benchmark.
  * Updated to easy-websocket 0.2, and pulled in hybi parser from there.
  * Documentation for client-side usage (fixes GH-7).
  * Fix typo in 'detroy' -> 'destroy'
  * Added travis CI
  * Fixed example.

0.1.0 / 2011-11-14
==================

  * Speed improvemements.
    * Faster socket initialization.
    * Faster parser in the order of 17%-40%
  * Added comprehensive tests.
  * Added benchmarks.


lib/protocols/drafts.js

/*!
 * socket.io-node
 * Copyright(c) 2011 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module requirements.
 */

var Socket = require('../socket')
  , EventEmitter = require('events')
  , crypto = require('crypto')
  , debug = require('debug')('wsio:drafts')

/**
 * Export the constructor.
 */

module.exports = WebSocket;
module.exports.Parser = Parser;

/**
 * HTTP interface constructor. Interface compatible with all transports that
 * depend on request-response cycles.
 *
 * @api public
 */

function WebSocket (req) {
  // parser
  var self = this;

  this.parser = new Parser();
  this.parser
    .on('data', function (packet) {
      debug(self.name + ' received data packet', packet);
      self.onMessage(packet);
    })
    .on('close', function () {
      self.end();
    })
    .on('error', function () {
      self.end();
    })

  Socket.call(this, req);
};

/**
 * Inherits from Socket.
 */

WebSocket.prototype.__proto__ = Socket.prototype;

/**
 * Websocket identifier
 *
 * @api public
 */

WebSocket.prototype.name = 'websocket-hixie';

/**
 * Websocket draft version
 *
 * @api public
 */

WebSocket.prototype.protocolVersion = 'hixie-76';

/**
 * Called when the socket connects.
 *
 * @api private
 */

WebSocket.prototype.onOpen = function () {
  var self = this;

  this.socket.setNoDelay(true);

  this.buffer = true;
  this.buffered = [];

  if (this.req.headers.upgrade !== 'WebSocket') {
    debug(this.name + ' connection invalid');
    this.end();
    return;
  }

  var origin = this.req.headers.origin
    , location = ((this.req.headers['x-forwarded-proto'] === 'https' || this.socket.encrypted) ? 'wss' : 'ws')
               + '://' + this.req.headers.host + this.req.url
    , waitingForNonce = false;

  if (this.req.headers['sec-websocket-key1']) {
    // If we don't have the nonce yet, wait for it (HAProxy compatibility).
    if (! (this.req.head && this.req.head.length >= 8)) {
      waitingForNonce = true;
    }

    var headers = [
        'HTTP/1.1 101 WebSocket Protocol Handshake'
      , 'Upgrade: WebSocket'
      , 'Connection: Upgrade'
      , 'Sec-WebSocket-Origin: ' + origin
      , 'Sec-WebSocket-Location: ' + location
    ];

    if (this.req.headers['sec-websocket-protocol']){
      headers.push('Sec-WebSocket-Protocol: '
          + this.req.headers['sec-websocket-protocol']);
    }
  } else {
    this.protocolVersion = 'hixie-75';
    var headers = [
        'HTTP/1.1 101 Web Socket Protocol Handshake'
      , 'Upgrade: WebSocket'
      , 'Connection: Upgrade'
      , 'WebSocket-Origin: ' + origin
      , 'WebSocket-Location: ' + location
    ];
  }

  this.emit('headers', headers);

  try {
    this.socket.write(headers.concat('', '').join('\r\n'));
    this.socket.setTimeout(0);
    this.socket.setNoDelay(true);
    this.socket.setEncoding('utf8');
  } catch (e) {
    this.end();
    return;
  }

  if (waitingForNonce) {
    this.socket.setEncoding('binary');
  } else if (this.proveReception(headers)) {
    process.nextTick(function() {
      if ('opening' == self.readyState) {
        self.readyState = 'open';
        self.emit('open');
        self.flush();
      }
    });
  }

  var headBuffer = '';

  this.socket.on('data', function (data) {
    if (waitingForNonce) {
      headBuffer += data;

      if (headBuffer.length < 8) {
        return;
      }

      // Restore the connection to utf8 encoding after receiving the nonce
      self.socket.setEncoding('utf8');
      waitingForNonce = false;

      // Stuff the nonce into the location where it's expected to be
      self.req.head = headBuffer.substr(0, 8);
      headBuffer = '';

      if (self.proveReception(headers)) {
         process.nextTick(function() {
          if ('opening' == self.readyState) {
            self.readyState = 'open';
            self.emit('open');
            self.flush();
          }
        });
      }

      return;
    }

    self.parser.add(data);
  });
};

/**
 * Writes to the socket.
 *
 * @api private
 */

WebSocket.prototype.write = function (data) {
  if ('open' == this.readyState && 'open' == this.socket.readyState) {
    this.drained = false;

    if (this.buffer) {
      this.buffered.push(data);
      return this;
    }

    var length = Buffer.byteLength(data)
      , buffer = new Buffer(2 + length);

    buffer.write('\x00', 'binary');
    buffer.write(data, 1, 'utf8');
    buffer.write('\xff', 1 + length, 'binary');

    try {
      if (this.socket.write(buffer)) {
        this.drained = true;
      }
    } catch (e) {
      this.end();
    }

    debug(this.name + ' writing', data);
  }
};

/**
 * Flushes the internal buffer
 *
 * @api private
 */

WebSocket.prototype.flush = function () {
  this.buffer = false;

  for (var i = 0, l = this.buffered.length; i < l; i++) {
    this.write(this.buffered.splice(0, 1)[0]);
  }
};

/**
 * Finishes the handshake.
 *
 * @api private
 */

WebSocket.prototype.proveReception = function (headers) {
  var self = this
    , k1 = this.req.headers['sec-websocket-key1']
    , k2 = this.req.headers['sec-websocket-key2'];

  if (k1 && k2){
    var md5 = crypto.createHash('md5');

    [k1, k2].forEach(function (k) {
      var n = parseInt(k.replace(/[^\d]/g, ''))
        , spaces = k.replace(/[^ ]/g, '').length;

      if (spaces === 0 || n % spaces !== 0){
        debug('Invalid ' + self.name + ' key: "' + k + '".');
        self.end();
        return false;
      }

      n /= spaces;

      md5.update(String.fromCharCode(
        n >> 24 & 0xFF,
        n >> 16 & 0xFF,
        n >> 8  & 0xFF,
        n       & 0xFF));
    });

    md5.update(this.req.head.toString('binary'));

    if ('open' != this.socket.readyState) {
      this.end();
      return false;
    }

    try {
      this.socket.write(md5.digest('binary'), 'binary');
    } catch (e) {
      this.end();
    }
  }

  return true;
};

/**
 * Writes a payload.
 *
 * @api private
 */

WebSocket.prototype.payload = function (msgs) {
  for (var i = 0, l = msgs.length; i < l; i++) {
    this.write(msgs[i]);
  }

  return this;
};

/**
 * WebSocket parser
 *
 * @api public
 */

function Parser () {
  this.buffer = '';
  this.i = 0;
};

/**
 * Inherits from EventEmitter.
 */

Parser.prototype.__proto__ = EventEmitter.prototype;

/**
 * Adds data to the buffer.
 *
 * @api public
 */

Parser.prototype.add = function (data) {
  this.buffer += data;
  this.parse();
};

/**
 * Parses the buffer.
 *
 * @api private
 */

Parser.prototype.parse = function () {
  for (var i = this.i, chr, l = this.buffer.length; i < l; i++){
    chr = this.buffer[i];

    if (this.buffer.length == 2 && this.buffer[1] == '\u0000') {
      this.emit('close');
      this.buffer = '';
      this.i = 0;
      return;
    }

    if (i === 0) {
      if (chr != '\u0000') {
        this.error('Bad framing. Expected null byte as first frame');
      } else {
        continue;
      }
    }

    if (chr == '\ufffd') {
      this.emit('data', this.buffer.substr(1, i - 1));
      this.buffer = this.buffer.substr(i + 1);
      this.i = 0;
      return this.parse();
    }
  }
};

/**
 * Handles an error
 *
 * @api private
 */

Parser.prototype.error = function (reason) {
  this.buffer = '';
  this.i = 0;
  this.emit('error', reason);
  return this;
};
